File: test-suite/src/test/java/io/micronaut/security/docs/blockingauthenticationprovider/AuthenticationProviderTest.java
package io.micronaut.security.docs.blockingauthenticationprovider;

import io.micronaut.context.annotation.Property;
import io.micronaut.context.annotation.Requires;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.client.BlockingHttpClient;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.http.client.exceptions.HttpClientResponseException;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.rules.SecurityRule;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import java.util.Collections;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

@Property(name = "spec.name", value = "AuthenticationProviderTest")
@MicronautTest
class AuthenticationProviderTest {

    @Test
    void authProvider(@Client("/") HttpClient httpClient) {
        BlockingHttpClient client = httpClient.toBlocking();
        String json = assertDoesNotThrow(() -> client.retrieve(createRequest("user", "password")));
        String expected = """
                {"message":"Hello World"}""";
        assertEquals(expected, json);
        HttpRequest<?> request = createRequest("user", "wrong");
        Executable e = () -> client.retrieve(request);
        HttpClientResponseException ex = assertThrows(HttpClientResponseException.class, e);
        assertEquals(HttpStatus.UNAUTHORIZED, ex.getStatus());
    }

    private HttpRequest<?> createRequest(String userName, String password) {
        return HttpRequest.GET("/messages").basicAuth(userName, password);
    }

    @Requires(property = "spec.name", value = "AuthenticationProviderTest")
    @Controller("/messages")
    static class HelloWorldController {

        @Secured(SecurityRule.IS_AUTHENTICATED)
        @Get
        Map<String, Object> index() {
            return Collections.singletonMap("message", "Hello World");
        }
    }
}

File: test-suite-python/src/test/python/micronaut/security/docs/blockingauthenticationprovider/AuthenticationProviderTest.py
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property, Requires
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.annotation import Controller, Get
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.http.client.exceptions import HttpClientResponseException
from micronaut.security.annotation import Secured
from micronaut.security.rules import SecurityRule
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test


@Property(name="spec.name", value="AuthenticationProviderTest")
@MicronautTest
class AuthenticationProviderTest:
    httpClient: Annotated[HttpClient, Inject, Client("/")]

    @Test
    def test_auth_provider(self):
        client = self.httpClient.toBlocking()
        json = client.retrieve(self.create_request("user", "password"))
        assert json == '{"message":"Hello World"}'
        try:
            client.retrieve(self.create_request("user", "wrong"))
            assert False, "HttpClientResponseException expected"
        except HttpClientResponseException as e:
            assert e.getStatus() == HttpStatus.UNAUTHORIZED

    def create_request(self, user_name: str, password: str):
        return HttpRequest.GET("/messages").basicAuth(user_name, password)


@Requires(property="spec.name", value="AuthenticationProviderTest")
@Controller("/messages")
class HelloWorldController:

    @Secured(SecurityRule.IS_AUTHENTICATED)
    @Get
    def index(self) -> dict:
        return {"message": "Hello World"}
